### PR TITLE
Version 2.0 [WIP]

### PR DESCRIPTION
**Pull request by emadum**
- Refactor to Typescript (in progress)
- Switch to IORedis (todo)
- Improve UMF messaging (todo)
  - Consistently use short form
  - Support binary message packing
- Refactor to use async/await (todo)

### CODE REVIEW

File: index.ts [hunk 1]
```diff
@@ -0,0 +1,1946 @@
...
+    }
+    const initPromise = this.initPromise(config);
+    this.ready = () => initPromise;
+    return initPromise;
+  }
+  private initPromise(config: Types.ServiceConfig) {
+    return new Promise((resolve, reject) => {
+      let loader = (newConfig) => {
+        return Promise.series(this.registeredPlugins, (plugin) => plugin.setConfig(newConfig.hydra))
+          .then((..._results) => {
```
Comment (cjus) [inline index.ts]: It would be nice to see Async/Await where ever possible.

File: index.ts [hunk 2]
```diff
@@ -0,0 +1,1946 @@
...
+  }
+
+  /**
+   * @name _connectToRedis
+   * @summary Configure access to Redis and monitor emitted events.
+   * @private
+   * @param {object} config - Redis client configuration
+   * @return {object} promise - resolves or reject
+   */
+  _connectToRedis(config: Types.IRedisOptions): Promise<Types.IRedis> {
```
Comment (cjus) [inline index.ts]: should cleanup code to remove _ prefix and just use private / protected.

File: index.ts [hunk 3]
```diff
@@ -0,0 +1,1946 @@
...
+        this.redisdb.keys(pattern, (err, result) => {
+          if (err) {
+            resolve([]);
+          } else {
+            resolve(result);
+          }
+        });
+      } else {
+        let doScan = (cursor, pattern, retSet) => {
+          this.redisdb.scan(cursor, 'MATCH', pattern, 'COUNT', KEYS_PER_SCAN, (err, result) => {
```
Comment (cjus) [inline index.ts]: the interface to redis should return promises and then just use async/await for cleaner code.

File: types.ts [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+export type ServiceConfig = {
+  environment: string;
+  hydra: HydraConfig;
+};
+export type ServiceDetails = {
+  serviceName: string;
+  serviceVersion?: string;
+  serviceIP?: string;
+  serviceDNS?: string;
+  servicePort?: any;
```
Comment (cjus) [inline types.ts]: why any?

File: index.ts (hunk 3 above)
Comment (emadum) [reply index.ts]: Yeah, the interface between `node_redis` and `ioredis` is slightly different - I'd like to abstract that away as you described.

File: types.ts (hunk 4 above)
Comment (emadum) [reply types.ts]: I was having some difficulty getting it to compile with a stricter requirement - I'll take another look now that I'm a little more familiar with TS, a union type will probably do the trick.

**Comment by ricardovf** [comment]
Is there a date set to finish this new version? Im considering using this project (seems really nice), but im a little worried if it's really alive and going to have bugs fixed and new features. Thanks!

**Pull request closed without merging**